### PR TITLE
chore: refactor dictionary handling

### PR DESCRIPTION
**Pull request by ignatiusmb**
Moves dictionary files into a single object in a file, this should reduce bundle size and slightly improve performance because we'll be directly importing the dictionary object, skipping the middleman, so to speak.

I tried to reuse some of the types, but it seems some of them are unused or not quite up to date with the current code, so I updated them as well and tried to match it with the current code as much as possible.

### CODE REVIEW

**Review by robertsLando — APPROVED** [review]

**Review by robertsLando — APPROVED** [review]

**Comment by robertsLando** [comment]
cc @jesec

**Comment by robertsLando** [comment]
@ignatiusmb WOuld like to know @jesec opinion on this before merging

**Comment by ignatiusmb** [comment]
Another pair of eyes never goes wrong 😄

**Comment by ignatiusmb** [comment]
What's left to do?

**Review by robertsLando — APPROVED** [review]

File: test/test-77-compare-dicts-and-tests/main.js
Comment (baparham) [inline test/test-77-compare-dicts-and-tests/main.js:9]: presumably this all of the sudden needs to require the `lib-es5` version because it's the transpiled javascript, not the direct javascript it was previously, right?

File: test/test-77-compare-dicts-and-tests/main.js
Comment (baparham) [inline test/test-77-compare-dicts-and-tests/main.js:11]: There's not really an advantage here to inlining the readdir and the filter, I'd rather you extract the filter function into a named function defining what the filter is actually doing here, like `excludeIsolatorTests` or something

File: lib/walker.ts
Comment (baparham) [inline lib/walker.ts:696]: why do you need to cast as a string here? I think that may be a symptom of an earlier typing problem that should be resolved as part of this refactor.

File: lib/walker.ts
Comment (baparham) [inline lib/walker.ts:694]: while you are refactoring, please go ahead and rename variables to meaningful names, like `escapedRegEx` in this case.

File: lib/walker.ts
Comment (baparham) [reply lib/walker.ts:696]: I also can't see how this doesn't hit an array out of bounds exception. The current `for` loop logic is convoluted and should probably be refactored to be readable to future contributors. What is it doing? Can parts be broken out into meaningful function names?

**Review by baparham — COMMENTED** [review]
I'm on the fence on this one, can you clarify what problem you are trying to solve? This is obviously a refactor, but there are still things I'd like to see improved while you are in here so that it's not just "different and not understandable" but "different and _more_ understandable" to future contributors.

on one hand, it is nice to have the modules in their own separate files, since they do not relate to one another. Changes in one module file should not be part of the history when looking at another module. This pulls them all into one mega-file. Granted, they don't change often, so churn should be low.

On the other hand, the dictionary parsing code got much simpler, and in theory, more maintainable, which is a good thing.

**Comment by baparham** [comment]
The "not understandable" reference was to existing code and variable naming, which you (rightly) continued to follow, not a comment on your code specifically. I'm just looking for the code quality to improve more while you are touching these parts of the code.

File: test/test-77-compare-dicts-and-tests/main.js
Comment (ignatiusmb) [reply test/test-77-compare-dicts-and-tests/main.js:11]: I don't see any benefit from doing that and I don't think we should touch anything more than what this PR does, especially with old tests like this.

**Comment by ignatiusmb** [comment]
> I'm just looking for the code quality to improve more while you are touching these parts of the code.

I would love to do that, but the changes here are already huge as it is, I would prefer to have that in a separate PR as it may get out of scope if we start touching a lot of unrelated code here.

File: lib/walker.ts
Comment (ignatiusmb) [reply lib/walker.ts]: I reapproached this refactor from the types instead, it should clarify the intent better now.

**Comment by ignatiusmb** [comment]
> can you clarify what problem you are trying to solve? [...] on one hand, it is nice to have the modules in their own separate files, since they do not relate to one another. Changes in one module file should not be part of the history when looking at another module. This pulls them all into one mega-file. Granted, they don't change often, so churn should be low.

To keep it simple and consistent in one file and type-checked with TypeScript, make it easier to add or amend patches in the future, and reduce inefficient `path` concatenation and `fs` calls.

**Comment by ignatiusmb** [comment]
Test failures seems unrelated?

File: lib/walker.ts
Comment (ignatiusmb) [reply lib/walker.ts]: Didn't think about it on the moment but refactoring this in any way from the types or changing how the loop works will be a breaking change to the `"patches"` API.

Comment (baparham) [reply lib/walker.ts]: it seems like you've made it better with your most recent changes though. I do not know what the existing community does with these dictionary entries, if they tend to patch them themselves (in which case, yes, this would break their patches) or if they are just used by pkg itself and can be considered an implementation detail.

I'd argue that these are the latter, and as an implementation detail, can't be relied upon by anyone patching them themselves at runtime to have some sort of stable API that we'd have to semver control.

Do you have any hints about what the docs say about these dictionary entries?

File: test/test-77-compare-dicts-and-tests/main.js
Comment (baparham) [reply test/test-77-compare-dicts-and-tests/main.js:11]: I meant if you are going to change these lines anyway, why not answer some of the questions posed by the obscure `.filter` logic...what is that filtering doing? simply breaking that filter function into a named function that actually tells the reader what it's doing, or why would help the next person come in and also continue to improve the code next time they touch this line of code.

File: lib/walker.ts
Comment (baparham) [reply lib/walker.ts]: I can't find any hint that these dictionary entries should be messed with by users. no "do this to support library XYZ" at all.

Comment (ignatiusmb) [reply lib/walker.ts]: These are the only "docs" I could find regarding the patches https://github.com/vercel/pkg/wiki/Developers#how-it-works, and that's for the developers of pkg (or more accurately, maintainers). I think you might be right that these are only used internally and by pkg itself, so I'm guessing it's okay to move forward with the recent changes?

File: test/test-77-compare-dicts-and-tests/main.js
Comment (ignatiusmb) [reply test/test-77-compare-dicts-and-tests/main.js]: I see, I've pushed some changes for this file, did I understand you correctly with those changes?

File: lib/walker.ts
Comment (baparham) [reply lib/walker.ts]: I think this conversation can be resolved then, thanks for the discussion and thought

**Review by baparham — APPROVED** [review]
My initial concerns have been addressed, so I see no reason this shouldn't go in. Thanks for the patience @ignatiusmb

**Review by robertsLando — CHANGES_REQUESTED** [review]
@ignatiusmb Could you resolve conflicts with master please?

File: lib/walker.ts
Comment (robertsLando) [inline lib/walker.ts:676]: Wait this is a breaking change... There are users that provide dictionary from package.json, in case them were using the old replace format this would stop working, could you add a fallback to also make replace work as before?

File: test/test-50-package-json-3/package.json
Comment (robertsLando) [inline test/test-50-package-json-3/package.json:10]: Like this case

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Pull request closed without merging**
